rfc: Explicit mechanism for working with Temporary Files

apparentlymart opened the pull request:
This document proposes a new referenceable symbol, path.temp, which returns a module-instance-specific directory where temporary files can be written.

Those files can then be assumed to persist at the same paths from the plan phase to the apply phase, even if those phases are run in different working directories on different computers.

The primary motivation for suggesting this _now_ is to try to start down a road towards deprecating the existing pattern of writing temporary files into `path.module`, which would make  it more likely that we would be able to treat more module packages as immutable in future for reasons like #1086 and #1942.

However, I believe the feature is also useful in its own right as a more ergonomic approach to a somewhat-common need.

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

cam72cam commented on rfc/20241003-temporary-files.md:
```diff
@@ -0,0 +1,295 @@
+
+- Use `path.module` to generate a path relative to a specific module, and then have that module return its generated path as an output value so that other parts of the configuration can read the generated file.
+
+    This is largely the same as the first point but has the minor benefit of allowing each module to manage its own filesystem namespace. However, it's particularly problematic when a module is called using `for_each` or `count` because then multiple instances of the module all share a single directory, and will clobber each other's written files unless specialc care is taken.
```
Typo: `specialc`.

I've got https://github.com/opentofu/opentofu/issues/1086 assigned to me to remind me to update the documentation with this fact / warning.

cam72cam commented:
I definitely see the benefit of this RFC to move us in the direction toward immutable modules.

I've got some questions on implementation and workflows, this is my initial brain dump from reading it through:

* Who manages the `.terraform/temp` directory?
  * If the each plan command produces it's own temporary directory, with the idea that multiple temporary directories may coexist and need unique plan IDs, when are they deleted?
  * Instead of storing the temporary files in a location with a unique plan ID, why not simply clean the temp directory after plan has completed?  This would also solve the issue of including the temp directory in the plan file, though it would require additional work to re-generate the file during the apply.  It's a question of which problems we are trying to solve.
* What problem are people using path.module trying to solve?
  * Downloading a specific asset to a unique path for caching purposes?  Compiling a binary for use somewhere to save time on subsequent runs?  Generating a large file locally based on data unique to the plan invocation?
  * Is there a difference between path.temp and path.cache?  Do common use-cases all fit under a single umbrella or do we need both concepts?
    * If so, should path.cache be included in the plan file?
* Is the name path.temp too generic?  As someone first encountering this concept, my first instinct is that it is a global temporary directory for tofu to write to and is not module specific.
* Why hash the module path? Potential windows path length issues? Preventing module authors from guessing relative paths to jump into sister instances/modules?

I also had a though related to the deprecation of path.module and it's impact on immutable modules.  Would it be difficult / time expensive to parse the contents of a module and look for usages of paths.module to determine if it is "safe" to dedup?  My initial thought is that we don't have the full schema during init, but could potential rely on AST or similar level inspection for that particular lookup?  This however does not handle scenarios where modules are behaving and simply reading from files shipped alongside the module.

apparentlymart commented:
Thanks for those questions! I will update the RFC to answer most of them today, and then also post comments here to answer the individual questions more directly.

The one I'm not really sure how to answer directly myself is the question of what it should be named, since of course that is a highly subjective question. I would be curious to learn how others are thinking about the tradeoff between brevity (what I was aiming for, since I wanted something of similar visual shape and length as the other `path.` symbols) vs. clarity, and if there are any concrete suggestions for names that might better balance those two priorities.

apparentlymart commented:
Hello again! As promised, I've pushed a new draft which contains answers to the questions asked so far. I'm also going to try to share similar information Q&A-style here just so it's easier to follow this as a conversation rather than try to guess what each of my markdown edits was intended to respond to. If you'd prefer to interact with the RFC text rather than the side-conversation then you could also [review the diff from yesterday's draft](https://github.com/opentofu/opentofu/pull/2049/commits/cb11021582da8c865a2f03d69eecdc41ecadd2e8).

> I would like to offer for consideration that we should discuss if we would like to support sandboxing in any way, shape or form in the future. [...]

I had actually intended to touch on this more in my first draft but after writing all day I forgot to return and elaborate on the passing remark I'd made about it. :grinning: 

I've added [Provider Sandboxing](https://github.com/opentofu/opentofu/blob/rfc-temporary-files/rfc/20241003-temporary-files.md#provider-sandboxing) as a new section under "Future Considerations" where I discussed my thoughts on this in more detail, but the summary is that I don't think this proposal makes a future sandboxing effort considerably _harder_, though it also doesn't necessarily make it any easier.

I think there is at least some chance that `path.temp` is widely adopted, and the assumption of the entire configuration tree being mutable becomes "legacy", in which case the hypothetical future sandboxing proposal would need only to carve some way to share the _temporary directories_ between the sandboxed components, rather than the entire filesystem. However, I can't possibly predict whether `path.temp` will be compelling enough to motivate retrofitting into existing modules; the sandboxing proposal might need to answer for the current `path.module` and `path.cwd` patterns regardless of whether we introduce `path.temp`.

> Who manages the `.terraform/temp` directory? 

That's a good question and I've added it to "Open Questions" for now, since I'm curious to hear what others thing about it before I commit to an answer. However, my "open question" does include one possible answer for how this could work, aiming to find a compromise between automatic cleanup in the common case vs. making it easier to debug when something goes wrong.

> What problem are people using `path.module` trying to solve?

To try to start answering this I've added an appendix [Use-case Examples](https://github.com/opentofu/opentofu/blob/rfc-temporary-files/rfc/20241003-temporary-files.md#appendix-use-case-examples) where I wrote out the two main applications of the `path.module`/`path.cwd` pattern that I've seen in broad use by various different users.

I must admit I'm not entirely sure what you are implying by the hypothetical `path.cache`, but just from the name you chose I'm guessing you are thinking about a location that would intentionally persist from one plan/apply round to the next for use only as an optimization, rather than as a crucial part of the behavior. Is that right?

If so, I don't personally have any concrete examples of that, but if you have some then I'd be happy to consider how they might fit. I've intentionally framed this proposal so far as _explicitly_ for temporary files because a common hazard I've heard about in discussion with users is when they accidentally leave generated files on disk between rounds and that somehow causes the second round to plan a different set of actions than it ought to have with a clean working directory.

(I've not yet added anything specifically about `path.cache` to the proposal because I'd like to understand the idea better first, but I'm happy to work that into another draft if we find some compelling examples to discuss either as additional requirements for this proposal, or to be explicit about these use-cases being out of scope.)

> Is the name `path.temp` too generic?

As I mentioned previously, this is a highly subjective question and therefore not something I can directly answer, but I did add some additional language to try to justify why _to me_ `path.temp` reads as "the temporary directory for this module instance" rather than "the global temporary directory for the whole system".

A short version: the "references to named values" mechanism explicitly offers a separate namespace for each module instance, and so symbols that have global scope are the exception rather than the rule. I read all of them as effectively implying "[whatever the definition is] _for this module instance_" unless the documentation describes it differently.

However, I did primarily choose it because it is short and concise, and so the argument above is a retroactive justification. I'm still happy to consider other possibilities.

> Why hash the module path?

I've written longer words about this inline, but the main points are:

- Predictable length, regardless of the depth of module call nesting. This is in response to the goal of keeping the generated path strings as compact as possible.
- Predictable character set, whereas a plain module instance address can contain literally any Unicode character due to `for_each` instance keys. This one comes from prior experience with significant differences in filename treatment across operating systems and filesystems: the acceptable characters, the normalization rules, and the case folding rules all differ in subtle and not-so-subtle ways on different systems, so trying to generate filenames with predictable uniqueness when using the entire Unicode repertoire is unreasonably complicated.

> [...] Would it be difficult / time expensive to parse the contents of a module and look for usages of paths.module to determine if it is "safe" to dedup? [...]

This one is hard to answer exhaustively with the thinking I've done so far, but the main hazard I want to draw attention to in thinking about this (which I've also described in longer words in the RFC) is that the unit of module installation is _module packages_, while the unit of module calling at runtime is individual _modules_.

Therefore if we intend to try to share a single local directory among all calls into a single module package then we need to make sure to consider all of the modules from that package that the current configuration is using across all entry points. That sort of analysis during module installation has no precedent today, but I think it's possible in principle.

However, I find myself more strongly drawn to the other idea of introducing a new place for entire-module-package metadata, because that seems like a clearer and more direct solution: although it does require the creator of a module package to include something extra in it, that extra thing is hopefully easier to understand the implications of. As noted inline, I've also encountered potential use-cases for module package metadata numerous times before -- though I admittedly don't have concrete examples directly to hand right now -- so I have reason to believe that this would be an investment that we could make further use of in future to solve other module-installation-related problems.

apparentlymart commented on rfc/20241003-temporary-files.md:
This item should start with "The _module_ registry protocol". (This point is trying to point out a significant difference in the trust/security model of module registries vs. provider registries.)

apparentlymart commented on rfc/20241003-temporary-files.md:
```suggestion
    Each of those installation methods adds some additional variation into proceedings. For example, installing directly from a Git repository causes the resulting package directory to contain a `.git` directory containing repository objects and metadata, and so a naive checksum calculation would produce a different result than if identical module package content had been installed from a `.zip` archive in an S3 bucket. Normalizing those differences away is possible with some care, but there are far more different cases to cover than as for providers.
```

apparentlymart commented on rfc/20241003-temporary-files.md:
```suggestion
- Whereas OpenTofu requires an entire configuration to agree on a single version to use for each provider, currently each `module` block is able to make its own independent selection about which version of a module to use.
```

apparentlymart commented on rfc/20241003-temporary-files.md:
```suggestion
Before beginning the true apply phase in the language runtime, Tofu CLI must ensure that the directory structure under the generated prefix matches the files present under `tmp/` in the plan file, which includes creating any files that aren't already present, updating any pre-existing files to match the plan file, and deleting any files that are not present in the plan file. The ultimate goal is to create the illusion that the plan is being applied in the same filesystem where it was created, so that file paths that were captured as strings in the OpenTofu execution plan
```

apparentlymart commented:
I've marked this as being a draft because I have some further edits to deal with (represented by the self-comments above) but this is not currently my focus so I will not be able to spend time on that in the near future.

I'll revisit this again at a later date and hopefully get this back into a reviewable state.

joeaawad commented on rfc/20241003-temporary-files.md:
```diff
@@ -0,0 +1,447 @@
+
+OpenTofu has no built-in features for managing temporary files, and so informal patterns have emerged. The two most prominent approaches are:
+
+- Use `path.cwd` to generate a path referring to a file in the current working directory, and then pass that absolute path to all of the providers that need to write or read the file in question.
```
Worth calling out that there is also `path.root` and per https://opentofu.org/docs/language/expressions/references/#filesystem-and-workspace-info it is recommended over `path.cwd`
